src/tasks/models/manager.ts

```diff
@@ -153,59 +206,83 @@ export class TaskManager {
+      }
 
-    const previousStatus = task.status;
+      let nextStatus: TaskOperationStatus = status;
+      let dataToUpdate = undefined;
 
-    if (nextStatus === TaskOperationStatus.FAILED) {
-      nextStatus = task.attempts < task.maxAttempts ? TaskOperationStatus.RETRIED : TaskOperationStatus.FAILED;
-      dataToUpdate = nextStatus === TaskOperationStatus.FAILED ? {} : { attempts: task.attempts + 1 };
-    }
+      const previousStatus = task.status;
 
-    const newPersistedSnapshot = updateTaskMachineState(nextStatus, task.xstate);
+      if (nextStatus === TaskOperationStatus.FAILED) {
+        nextStatus = task.attempts < task.maxAttempts ? TaskOperationStatus.RETRIED : TaskOperationStatus.FAILED;
+        dataToUpdate = nextStatus === TaskOperationStatus.FAILED ? {} : { attempts: task.attempts + 1 };
+      }
 
-    const updateQueryBody = {
-      where: {
-        id: taskId,
-      },
-      data: { ...dataToUpdate, status: nextStatus, xstate: newPersistedSnapshot },
-    };
+      const newPersistedSnapshot = updateTaskMachineState(nextStatus, task.xstate);
+
+      const updateQueryBody = {
+        where: {
+          id: taskId,
+        },
+        data: { ...dataToUpdate, status: nextStatus, xstate: newPersistedSnapshot },
+      };
+
+      // update task current status
+      // const updatedTask = await this.prisma.$transaction(async (tx) => {
+      const updatedTask = await tx.task.update(updateQueryBody);
 
-    // update task current status
-    await this.prisma.task.update(updateQueryBody);
+      const updateSummaryPayload: UpdateSummaryCount = {
+        add: { status: nextStatus, count: 1 },
+        remove: { status: previousStatus, count: 1 },
+      };
 
-    // update stage progress data
-    const stage = await this.stageManager.getStageById(task.stageId);
+      await this.stageManager.updateStageProgressFromTaskChanges(task.stageId, updateSummaryPayload, tx);
 
-    const updateSummaryPayload: UpdateSummaryCount = {
-      add: { status: nextStatus, count: 1 },
-      remove: { status: previousStatus, count: 1 },
-    };
+      return updatedTask; // Ensure the updated task is returned
+    });
 
-    await this.stageManager.updateStageProgressFromTaskChanges(task.stageId, stage.jobId, updateSummaryPayload);
+    return convertPrismaToTaskResponse(updatedTask);
+  }
+
+  /**
+   * Dequeues a task of the specified type with highest priority
+   * @param taskType - The type of task to dequeue
+   * @returns The dequeued task model with updated status
+   * @throws TaskNotFoundError when no suitable task is found
+   */
+  public async dequeue(taskType: TaskType): Promise<TaskModel> {
+    const queryBody = generatePrioritizedTaskQuery(taskType);
+
+    const task = await this.prisma.task.findFirst(queryBody);
+
+    if (!task) {
+      throw new TaskNotFoundError(tasksErrorMessages.taskNotFound);
+    }
+
+    const dequeuedTask = await this.updateStatus(task.id, TaskOperationStatus.IN_PROGRESS);
```

Comment:
i think theres still a problem here with race condition. need to think about it. maybe we need to add a check when updating that the status didnt change yet, and if it did prevent the update.